t/semaphore.t

```diff
@@ -0,0 +1,1068 @@
+                if not sem then
+                    ngx.log(ngx.ERR, err)
+                end
+                g.test = sem
+            end
+            local sem = g.test
+            local ok, err = sem:post()
+            if not ok then
+                ngx.log(ngx.ERR, err)
+            end
+    ';
+    content_by_lua '
+            ngx.print("post")
+            ngx.exit(200)
+    ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+200
+wait
+200
+post
+--- log_level: debug
+--- no_error_log
+[error]
+
+
+
+=== TEST 17: semaphore post in timer.at
+--- http_config eval: $::HttpConfig
+--- config
+    location /test{
+        content_by_lua '
+            local res1,res2 = ngx.location.capture_multi{
+                {"/sem_wait"},
+                {"/sem_post"},
+            }
+            ngx.say(res1.status)
+            ngx.say(res1.body)
+            ngx.say(res2.status)
+            ngx.say(res2.body)
+        ';
+    }
+
+    location /sem_wait {
+        content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local g = getmetatable(_G).__index
+            if not g.test then
+                local sem, err = semaphore.new(0)
+                if not sem then
+                    ngx.say(err)
+                    ngx.exit(500)
+                end
+                g.test = sem
+            end
+            local sem = g.test
+            local ok, err = sem:wait(2)
+            if ok then
+                ngx.print("wait")
+                ngx.exit(200)
+            else
+                ngx.exit(500)
+            end
+        ';
+    }
+
+    location /sem_post {
+    content_by_lua '
+            local g = getmetatable(_G).__index
+            local function func(premature, g)
+                local semaphore = require "ngx.semaphore"
+                if not g.test then
+                    local sem, err = semaphore.new(0)
+                    if not sem then
+                        ngx.log(ngx.ERR, err)
+                    end
+                    g.test = sem
+                end
+                local sem = g.test
+                local ok, err = sem:post()
+                if not ok then
+                    ngx.log(ngx.ERR, err)
+                end
+            end
+            ngx.timer.at(0, func,g)
+            ngx.sleep(2)
+            ngx.print("post")
+            ngx.exit(200)
+    ';
+    }
+--- no_check_leak
+--- request
+GET /test
+--- response_body
+200
+wait
+200
+post
+--- log_level: debug
+--- no_error_log
+[error]
+
+
+
+=== TEST 18: a light thread that to be killed is waitting a semaphore
+--- http_config eval: $::HttpConfig
+--- config
+    location /test {
+       content_by_lua '
+            local semaphore = require "ngx.semaphore"
+            local sem = semaphore.new(0)
+            if not sem then
+                error("create failed")
+            end
+            local function func(sem)
+                sem:wait(10)
```

Comment:
Add some output in this function to ensure this code path is actually run. The current checks are just too weak. Tests are code and can always have bugs. So better be strict and skeptical. The same applies to some other test cases in this test file.